File: src/main/java/com/dansoftware/boomega/gui/record/googlebook/GoogleBookConnectionView.java
package com.dansoftware.boomega.gui.record.googlebook;

import com.dansoftware.boomega.db.Database;
import com.dansoftware.boomega.db.data.Record;
import com.dansoftware.boomega.db.data.ServiceConnection;
import com.dansoftware.boomega.googlebooks.GoogleBooksQueryBuilder;
import com.dansoftware.boomega.googlebooks.SingleGoogleBookQuery;
import com.dansoftware.boomega.googlebooks.Volume;
import com.dansoftware.boomega.gui.context.Context;
import com.dansoftware.boomega.gui.googlebooks.GoogleBookDetailsPane;
import com.dansoftware.boomega.gui.googlebooks.SearchParameters;
import com.dansoftware.boomega.gui.googlebooks.join.GoogleBookJoinerOverlay;
import com.dansoftware.boomega.gui.record.show.RecordTable;
import com.dansoftware.boomega.gui.util.BaseFXUtils;
import com.dansoftware.boomega.gui.util.I18NButtonTypes;
import com.dansoftware.boomega.i18n.I18N;
import com.dansoftware.boomega.util.concurrent.CachedExecutor;
import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIcon;
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIconView;
import javafx.beans.property.ObjectProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;
import javafx.concurrent.Task;
import javafx.scene.Group;
import javafx.scene.Node;
import javafx.scene.control.*;
import javafx.scene.layout.Priority;
import javafx.scene.layout.StackPane;
import javafx.scene.layout.VBox;
import javafx.util.Duration;
import jfxtras.styles.jmetro.JMetroStyleClass;
import org.apache.commons.lang3.StringUtils;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Collections;
import java.util.List;
import java.util.Optional;
import java.util.concurrent.TimeUnit;
import java.util.stream.Collectors;

/**
 * A {@link GoogleBookConnectionView} is a panel that shows the connection between
 * a Google Book volume and a {@link Record}. It allows to remove, modify and create
 * connection.
 *
 * @author Daniel Gyorffy
 */
public class GoogleBookConnectionView extends VBox {

    private static final Logger logger = LoggerFactory.getLogger(GoogleBookConnectionView.class);

    private static final String STYLE_CLASS = "google-book-dock";

    private final Context context;
    private final Database database;
    private final Cache<String, DetailsPane> cache;

    private final ObjectProperty<Runnable> onRefreshed = new SimpleObjectProperty<>();
    private final ObjectProperty<Task<Volume>> lastTask = new SimpleObjectProperty<>();
    private final ObjectProperty<DetailsPane> currentDetailsPane = new SimpleObjectProperty<>();
    private final StringProperty currentGoogleHandle = new SimpleStringProperty() {
        @Override
        protected void invalidated() {
            final Task<Volume> task = lastTask.get();
            if (task != null && task.isRunning()) {
                logger.debug("Running task found");
                task.cancel();
                logger.debug("Task cancelled");
            }
        }
    };

    private List<Record> items;

    public GoogleBookConnectionView(@NotNull Context context,
                                    @NotNull Database database,
                                    @Nullable List<Record> items) {
        this.context = context;
        this.database = database;
        this.cache = buildCache();
        this.getStyleClass().add(STYLE_CLASS);
        this.setItems(items);
    }

    private Cache<String, DetailsPane> buildCache() {
        return Caffeine.newBuilder()
                .expireAfterWrite(1, TimeUnit.MINUTES)
                .build();
    }

    public void setOnRefreshed(@Nullable Runnable onRefreshed) {
        this.onRefreshed.set(onRefreshed);
    }

    public void setItems(@Nullable List<Record> items) {
        this.createCache();
        this.items = items;
        this.currentGoogleHandle.set(retrieveGoogleBookHandle(items));
        logger.debug("Google book handle retrieved: {}", currentGoogleHandle.get());
        buildBaseContent(currentGoogleHandle.get(), Optional.ofNullable(items).orElseGet(Collections::emptyList));
    }

    private void createCache() {
        final String handle = currentGoogleHandle.get();
        final DetailsPane pane = currentDetailsPane.get();
        if (handle != null && pane != null) {
            logger.debug("Creating cache for: '{}'...", currentGoogleHandle.get());
            cache.put(handle, pane);
            currentGoogleHandle.set(null);
            currentDetailsPane.set(null);
        }
    }

    private void refresh() {
        setItems(items);
        if (onRefreshed.get() != null) {
            onRefreshed.get().run();
        }
    }

    private void setContent(@NotNull Node content) {
        this.getChildren().setAll(content);
    }

    private void showProgress() {
        var progressBar = new ProgressBar();
        progressBar.setProgress(ProgressIndicator.INDETERMINATE_PROGRESS);
        getChildren().add(0, progressBar);
    }

    private void stopProgress() {
        getChildren().removeIf(e -> e instanceof ProgressBar);
    }

    private void buildBaseContent(@Nullable String googleHandle,
                                  @NotNull List<Record> items) {
        if (googleHandle != null) {
            buildDetailsPane(googleHandle);
        } else if (items.size() == 1) {
            setContent(new NoConnectionPlaceHolder(context, database, items.get(0), this::refresh));
        } else if (items.size() > 1) {
            setContent(new MultipleSelectionPlaceHolder());
        } else {
            setContent(new NoSelectionPlaceHolder());
        }
    }

    private void buildDetailsPane(String googleBookLink) {
        DetailsPane pane = cache.getIfPresent(googleBookLink);
        if (pane != null) {
            pane.items = items;
            currentDetailsPane.set(pane);
            setContent(pane);
        } else {
            Task<Volume> task = new VolumePullTask(googleBookLink);
            lastTask.set(task);
            CachedExecutor.INSTANCE.submit(task);
        }
    }

    @SuppressWarnings("ConstantConditions")
    @Nullable
    private String retrieveGoogleBookHandle(@Nullable List<Record> items) {
        if (items == null)
            return null;
        List<String> distinctHandles = items.stream()
                .map(Record::getServiceConnection)
                .map(ServiceConnection::getGoogleBookHandle)
                .distinct()
                .collect(Collectors.toList());
        if (distinctHandles.size() == 1)
            return distinctHandles.get(0);
        return null;
    }

    private class VolumePullTask extends Task<Volume> {

        private final String googleHandle;

        VolumePullTask(@NotNull String googleHandle) {
            this.googleHandle = googleHandle;
            this.initEventHandlers();
        }

        private void initEventHandlers() {
            this.setOnRunning(e -> showProgress());
            this.setOnSucceeded(event -> {
                logger.debug("Pull task succeeded.");
                stopProgress();
                currentDetailsPane.set(new DetailsPane(context, database, this.getValue(), items, GoogleBookConnectionView.this::refresh));
                setContent(currentDetailsPane.get());
            });
            this.setOnFailed(event -> {
                logger.error("Pull task failed.", event.getSource().getException());
                stopProgress();
                setContent(new ErrorPlaceHolder(context, event.getSource().getException()));
            });
        }

        @Override
        protected Volume call() throws Exception {
            logger.debug("Starting pull task...");
            return new SingleGoogleBookQuery(googleHandle).load();
        }
    }

    private static final class DetailsPane extends VBox {

        private final Context context;
        private final Database database;
        private final Runnable refresh;

        private List<Record> items;

        DetailsPane(@NotNull Context context,
                    @NotNull Database database,
                    @NotNull Volume volume,
                    @NotNull List<Record> items,
                    @NotNull Runnable refresh) {
            this.context = context;
            this.database = database;
            this.refresh = refresh;
            this.items = items;
            buildUI(volume);
        }

        private void buildUI(Volume volume) {
            setVgrow(this, Priority.ALWAYS);
            setSpacing(5.0);
            final var mainPane = new GoogleBookDetailsPane(context, volume);
            mainPane.setMinHeight(0);
            VBox.setVgrow(mainPane, Priority.ALWAYS);
            getChildren().add(mainPane);
            getChildren().add(buildRemoveButton());
        }

        private Button buildRemoveButton() {
            final var button = new Button(I18N.getValue("google.books.dock.remove_connection"));
            button.getStyleClass().add("remove-button");
            button.prefWidthProperty().bind(this.widthProperty());
            button.setOnAction(event -> {

                context.showDialog(
                        I18N.getValue("google.books.dock.remove.confirmation.title", items.size()),
                        buildPreviewTable(items),
                        it -> {
                            if (BaseFXUtils.typeEquals(it, ButtonType.YES)) {
                                CachedExecutor.INSTANCE.submit(buildConnectionRemoveTask());
                            }
                        }, I18NButtonTypes.CANCEL, I18NButtonTypes.YES);
            });
            return button;
        }

        private RecordTable buildPreviewTable(List<Record> items) {
            var recordTable = new RecordTable(0);
            recordTable.addColumn(RecordTable.ColumnType.INDEX_COLUMN);
            recordTable.addColumn(RecordTable.ColumnType.TYPE_INDICATOR_COLUMN);
            recordTable.addColumn(RecordTable.ColumnType.AUTHOR_COLUMN);
            recordTable.addColumn(RecordTable.ColumnType.TITLE_COLUMN);
            recordTable.getItems().setAll(items);
            recordTable.setPrefHeight(200);
            return recordTable;
        }

        private Task<Void> buildConnectionRemoveTask() {
            var task = new Task<Void>() {
                @SuppressWarnings("ConstantConditions")
                @Override
                protected Void call() {
                    items.stream()
                            .peek(record -> record.getServiceConnection().setGoogleBookLink(null))
                            .forEach(database::updateRecord);
                    return null;
                }
            };
            task.setOnRunning(event -> context.showIndeterminateProgress());
            task.setOnFailed(event -> {
                context.stopProgress();
                //TODO: ERROR DIALOG
                logger.error("Couldn't remove Google Book connection", event.getSource().getException());
                refresh.run();
            });
            task.setOnSucceeded(event -> {
                context.stopProgress();
                refresh.run();

                context.showInformationNotification(
                        I18N.getValue("google.books.dock.success_unjoin.title"),
                        null,
                        Duration.seconds(2)
                );
            });
            return task;
        }
    }

    private static final class ErrorPlaceHolder extends StackPane {

        private final Context context;
        private final Throwable cause;

        ErrorPlaceHolder(@NotNull Context context, Throwable cause) {
            this.context = context;
            this.cause = cause;
            buildUI();
        }

        private void buildUI() {
            this.getStyleClass().add(JMetroStyleClass.BACKGROUND);
            this.getChildren().add(
                    new Group(new VBox(20,
                            new StackPane(buildLabel()),
                            new StackPane(buildDetailsButton())))
            );
            VBox.setVgrow(this, Priority.ALWAYS);
        }

        private Label buildLabel() {
            var label = new Label(I18N.getValue("google.books.dock.placeholder.error"));
            label.getStyleClass().add("place-holder-label");
            return label;
        }

        private Button buildDetailsButton() {
            var button = new Button(
                    I18N.getValue("google.books.dock.placeholder.error.details"),
                    new MaterialDesignIconView(MaterialDesignIcon.DETAILS)
            );
            button.setOnAction(event -> context.showErrorDialog(StringUtils.EMPTY, StringUtils.EMPTY, (Exception) cause));
            return button;
        }
    }

    private static final class NoConnectionPlaceHolder extends StackPane {

        private final Context context;
        private final Database database;
        private final Record record;
        private final Runnable refresh;

        NoConnectionPlaceHolder(@NotNull Context context,
                                @NotNull Database database,
                                @NotNull Record record,
                                @NotNull Runnable refresh) {
            this.context = context;
            this.database = database;
            this.record = record;
            this.refresh = refresh;
            buildUI();
        }

        private void buildUI() {
            VBox.setVgrow(this, Priority.ALWAYS);
            this.getStyleClass().add(JMetroStyleClass.BACKGROUND);
            getChildren().add(buildContent());
        }

        private Node buildContent() {
            return new Group(new VBox(20,
                    new StackPane(buildLabel()),
                    buildConnectionButton()));
        }

        private Label buildLabel() {
            var label = new Label(I18N.getValue("google.books.dock.placeholder.noconn"));
            label.getStyleClass().add("place-holder-label");
            return label;
        }

        private Button buildConnectionButton() {
            var button = new Button(
                    I18N.getValue("google.books.dock.connection"),
                    new MaterialDesignIconView(MaterialDesignIcon.GOOGLE)
            );
            button.setOnAction(event -> showGoogleBookJoiner());
            return button;
        }

        public void showGoogleBookJoiner() {
            context.showOverlay(
                    new GoogleBookJoinerOverlay(context, buildSearchParameters(record), volume -> {
                        CachedExecutor.INSTANCE.submit(buildJoinActionTask(volume));
                    })
            );
        }

        private SearchParameters buildSearchParameters(@NotNull Record record) {
            return new SearchParameters()
                    .printType(record.getRecordType() == Record.Type.BOOK ?
                            GoogleBooksQueryBuilder.PrintType.BOOKS :
                            GoogleBooksQueryBuilder.PrintType.MAGAZINES)
                    .isbn(record.getIsbn())
                    .authors(String.join(",", Optional.ofNullable(record.getAuthors()).orElse(Collections.emptyList())))
                    .publisher(record.getPublisher())
                    .title(record.getTitle())
                    .language(record.getLanguage());
        }

        private Task<Void> buildJoinActionTask(Volume volume) {
            var task = new Task<Void>() {
                @SuppressWarnings("ConstantConditions")
                @Override
                protected Void call() {
                    record.getServiceConnection().setGoogleBookLink(volume.getSelfLink());
                    database.updateRecord(record);
                    return null;
                }
            };
            task.setOnRunning(event -> context.showIndeterminateProgress());
            task.setOnFailed(event -> {
                //TODO: ERROR DIALOG
                logger.error("Couldn't join with Google Book", event.getSource().getException());
                context.stopProgress();
            });
            task.setOnSucceeded(event -> {
                context.stopProgress();
                refresh.run();

                context.showInformationNotification(
                        I18N.getValue("google.books.dock.success_join.title"),
                        null,
                        Duration.seconds(2)
                );
            });
            return task;
        }
    }

    private static final class MultipleSelectionPlaceHolder extends StackPane {

        MultipleSelectionPlaceHolder() {
            buildUI();
        }

        private void buildUI() {
            this.getStyleClass().add(JMetroStyleClass.BACKGROUND);
            this.getChildren().add(buildLabel());
            VBox.setVgrow(this, Priority.ALWAYS);
        }

        private Label buildLabel() {
            var label = new Label(I18N.getValue("google.books.dock.placeholder.multiple"));
            label.getStyleClass().add("place-holder-label");
            return label;
        }
    }

    private static final class NoSelectionPlaceHolder extends StackPane {

        NoSelectionPlaceHolder() {
            buildUI();
        }

        private void buildUI() {
            this.getStyleClass().add(JMetroStyleClass.BACKGROUND);
            this.getChildren().add(buildLabel());
            VBox.setVgrow(this, Priority.ALWAYS);
        }

        private Label buildLabel() {
            var label = new Label(I18N.getValue("google.books.dock.placeholder.noselection"));
            label.getStyleClass().add("place-holder-label");
            return label;
        }
    }
}

File: src/main/java/com/dansoftware/boomega/gui/record/connection/GoogleBookConnectionView.kt
/*
 * Boomega
 * Copyright (C)  2021  Daniel Gyoerffy
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.dansoftware.boomega.gui.record.connection

import com.dansoftware.boomega.db.Database
import com.dansoftware.boomega.db.data.Record
import com.dansoftware.boomega.googlebooks.GoogleBooksQueryBuilder
import com.dansoftware.boomega.googlebooks.SingleGoogleBookQuery
import com.dansoftware.boomega.googlebooks.Volume
import com.dansoftware.boomega.gui.context.Context
import com.dansoftware.boomega.gui.googlebooks.GoogleBookDetailsPane
import com.dansoftware.boomega.gui.googlebooks.SearchParameters
import com.dansoftware.boomega.gui.googlebooks.join.GoogleBookJoinerOverlay
import com.dansoftware.boomega.gui.record.RecordTable
import com.dansoftware.boomega.gui.util.I18NButtonTypes
import com.dansoftware.boomega.gui.util.typeEquals
import com.dansoftware.boomega.i18n.I18N
import com.dansoftware.boomega.util.concurrent.CachedExecutor
import com.github.benmanes.caffeine.cache.Cache
import com.github.benmanes.caffeine.cache.Caffeine
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIcon
import de.jensd.fx.glyphs.materialdesignicons.MaterialDesignIconView
import javafx.concurrent.Task
import javafx.scene.Group
import javafx.scene.Node
import javafx.scene.control.*
import javafx.scene.layout.Priority
import javafx.scene.layout.StackPane
import javafx.scene.layout.VBox
import javafx.util.Duration
import jfxtras.styles.jmetro.JMetroStyleClass
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.concurrent.TimeUnit

/**
 * A [GoogleBookConnectionView] is a panel that shows the connection between
 * a Google Book volume and a [Record]. It allows to remove, modify and create
 * connection.
 *
 * @author Daniel Gyorffy
 */
class GoogleBookConnectionView(
    private val context: Context,
    private val database: Database,
    items: List<Record>
) : VBox() {

    private val detailsPane: GoogleBookDetailsPane = GoogleBookDetailsPane(context)
    private val volumeCache: Cache<String, Volume> = buildCache()

    var items: List<Record> = emptyList()
        set(value) {
            field = value
            handleNewItems(value)
        }

    var onRefreshed: Runnable? = null

    private var currentGoogleHandle: String? = null
    private var currentVolume: Volume?
        get() = detailsPane.volumeProperty().get()
        set(value) {
            detailsPane.volumeProperty().set(value)
        }

    private var lastTask: VolumePullTask? = null
        set(value) {
            field?.cancel()
            field = value
        }

    init {
        styleClass.add("google-book-dock")
        noSelectionUI()
        this.items = items
    }

    private fun buildCache(): Cache<String, Volume> =
        Caffeine.newBuilder()
            .expireAfterWrite(1, TimeUnit.MINUTES)
            .build()

    private fun handleNewItems(items: List<Record>) {
        retrieveGoogleBookHandle(items)?.let {
            normalUI()
            pullVolume(it)
        } ?: when {
            items.isEmpty() -> noSelectionUI()
            items.size == 1 -> noConnectionUI()
            else -> multipleSelectionUI()
        }
    }

    private fun pullVolume(handle: String) {
        VolumePullTask(handle).also { lastTask = it }.start()
    }

    private fun normalUI() {
        children.setAll(detailsPane, buildRemoveButton())
    }

    private fun noConnectionUI() {
        children.setAll(NoConnectionPlaceHolder())
    }

    private fun noSelectionUI() {
        children.setAll(NoSelectionPlaceHolder())
    }

    private fun multipleSelectionUI() {
        children.setAll(MultipleSelectionPlaceHolder())
    }

    private fun errorUI() {
        children.setAll(ErrorPlaceHolder())
    }

    private fun buildRemoveButton() =
        Button(I18N.getValue("google.books.dock.remove_connection")).apply {
            styleClass.add("remove-button")
            maxWidth = Double.MAX_VALUE
            VBox.setVgrow(this, Priority.ALWAYS)
            setOnAction {
                context.showDialog(
                    I18N.getValue("google.books.dock.remove.confirmation.title", items.size),
                    buildPreviewTable(items),
                    {
                        if (it.typeEquals(ButtonType.YES)) {
                            CachedExecutor.submit(ConnectionRemoveTask())
                        }
                    },
                    I18NButtonTypes.CANCEL,
                    I18NButtonTypes.YES
                )
            }
        }

    private fun buildPreviewTable(records: List<Record>) =
        RecordTable(0).apply {
            addColumn(RecordTable.ColumnType.INDEX_COLUMN)
            addColumn(RecordTable.ColumnType.TYPE_INDICATOR_COLUMN)
            addColumn(RecordTable.ColumnType.AUTHOR_COLUMN)
            addColumn(RecordTable.ColumnType.TITLE_COLUMN)
            items.setAll(records)
            prefHeight = 200.0
        }

    private fun retrieveGoogleBookHandle(items: List<Record>?): String? {
        return items?.map(Record::serviceConnection)
            ?.map { it?.googleBookHandle }
            ?.distinct()
            ?.singleOrNull()
    }

    private fun showProgress() {
        children.add(0, ProgressBar().apply {
            progress = ProgressIndicator.INDETERMINATE_PROGRESS
        })
    }

    private fun stopProgress() {
        children.removeIf { it is ProgressBar }
    }

    private inner class VolumePullTask(private val googleHandle: String) : Task<Volume?>() {

        init {
            initEventHandlers()
        }

        private fun initEventHandlers() {
            this.setOnRunning { showProgress() }
            this.setOnSucceeded {
                logger.debug("Pull task succeeded.")
                stopProgress()
                currentGoogleHandle = googleHandle
                currentVolume = value
            }
            this.setOnFailed {
                logger.error("Pull task failed.", it.source.exception)
                stopProgress()
                errorUI()
            }
        }

        override fun call(): Volume {
            logger.debug("Starting pull task...")
            return SingleGoogleBookQuery(googleHandle).load()
        }

        fun start() {
            cacheExisting()
            volumeCache.getIfPresent(googleHandle)?.let {
                logger.debug("Found cache for: '{}'", googleHandle)
                currentGoogleHandle = googleHandle
                currentVolume = it
            } ?: CachedExecutor.submit(this)
        }

        private fun cacheExisting() {
            currentGoogleHandle?.let {
                logger.debug("Creating cache for: '{}'...", it)
                volumeCache.put(it, currentVolume)
            }
        }
    }

    private inner class ConnectionRemoveTask() : Task<Unit>() {

        init {
            setOnRunning { context.showIndeterminateProgress() }
            setOnFailed {
                context.stopProgress()
                //TODO: error dialog
                logger.error("Couldn't remove Google Book connection", it.source.exception)
                onRefreshed?.run()
            }
            setOnSucceeded {
                context.stopProgress()
                onRefreshed?.run()
                noConnectionUI()
                context.showInformationNotification(
                    I18N.getValue("google.books.dock.success_unjoin.title"),
                    null,
                    Duration.seconds(2.0)
                )
            }
        }

        override fun call() {
            items.stream()
                .peek { it.serviceConnection?.googleBookHandle = null }
                .forEach(database::updateRecord)
        }
    }

    private class NoSelectionPlaceHolder() : StackPane() {
        init {
            buildUI()
        }

        private fun buildUI() {
            styleClass.add(JMetroStyleClass.BACKGROUND)
            children.add(buildLabel())
            setVgrow(this, Priority.ALWAYS)
        }

        private fun buildLabel() =
            Label(I18N.getValue("google.books.dock.placeholder.noselection")).apply {
                styleClass.add("place-holder-label")
            }
    }

    private class MultipleSelectionPlaceHolder : StackPane() {
        init {
            buildUI()
        }

        private fun buildUI() {
            styleClass.add(JMetroStyleClass.BACKGROUND)
            children.add(buildLabel())
            setVgrow(this, Priority.ALWAYS)
        }

        private fun buildLabel() =
            Label(I18N.getValue("google.books.dock.placeholder.multiple")).apply {
                styleClass.add("place-holder-label")
            }
    }

    private inner class ErrorPlaceHolder : StackPane() {
        init {
            styleClass.add(JMetroStyleClass.BACKGROUND)
            setVgrow(this, Priority.ALWAYS)
            buildUI()
        }

        private fun buildUI() {
            children.add(
                Group(
                    VBox(
                        20.0,
                        StackPane(buildLabel()),
                        StackPane(buildDetailsButton())
                    )
                )
            )
        }

        private fun buildLabel() =
            Label(I18N.getValue("google.books.dock.placeholder.error")).apply {
                styleClass.add("place-holder-label")
            }

        private fun buildDetailsButton() =
            Button(
                I18N.getValue("google.books.dock.placeholder.error.details"),
                MaterialDesignIconView(MaterialDesignIcon.DETAILS)
            ).apply {
                setOnAction {
                    context.showErrorDialog(
                        "",
                        "",
                        it.source as Exception
                    )
                }
            }
    }

    private inner class NoConnectionPlaceHolder() : StackPane() {

       init {
            styleClass.add(JMetroStyleClass.BACKGROUND)
            setVgrow(this, Priority.ALWAYS)
            buildUI()
        }

        private fun buildUI() {
            children.add(buildContent())
        }

        private fun buildContent(): Node {
            return Group(
                VBox(
                    20.0,
                    StackPane(buildLabel()),
                    buildConnectionButton()
                )
            )
        }

        private fun buildLabel() =
            Label(I18N.getValue("google.books.dock.placeholder.noconn")).apply {
                styleClass.add("place-holder-label")
            }

        private fun buildConnectionButton() = Button(
            I18N.getValue("google.books.dock.connection"),
            MaterialDesignIconView(MaterialDesignIcon.GOOGLE)
        ).apply {
            setOnAction { showGoogleBookJoiner() }
        }

        fun showGoogleBookJoiner() {
            context.showOverlay(
                GoogleBookJoinerOverlay(context, buildSearchParameters(items[0])) { volume ->
                    CachedExecutor.submit(
                        buildJoinActionTask(volume)
                    )
                }
            )
        }

        private fun buildSearchParameters(record: Record): SearchParameters {
            return SearchParameters()
                .printType(
                    if (record.recordType === Record.Type.BOOK)
                        GoogleBooksQueryBuilder.PrintType.BOOKS
                    else GoogleBooksQueryBuilder.PrintType.MAGAZINES
                )
                .isbn(record.isbn)
                .authors((record.authors ?: emptyList()).joinToString(","))
                .publisher(record.publisher)
                .title(record.title)
                .language(record.language)
        }

        private fun buildJoinActionTask(volume: Volume) =
            object : Task<Unit>() {

                init {
                    setOnRunning { context.showIndeterminateProgress() }
                    setOnFailed { event ->
                        //TODO: ERROR DIALOG
                        logger.error("Couldn't join with Google Book", event.source.exception)
                        context.stopProgress()
                    }
                    setOnSucceeded {
                        context.stopProgress()
                        onRefreshed?.run()
                        normalUI()
                        context.showInformationNotification(
                            I18N.getValue("google.books.dock.success_join.title"),
                            null,
                            Duration.seconds(2.0)
                        )
                    }
                }

                override fun call() {
                    items[0].serviceConnection?.googleBookHandle = volume.selfLink
                    database.updateRecord(items[0])
                }
            }

        init {
            buildUI()
        }
    }

    companion object {
        private val logger: Logger = LoggerFactory.getLogger(GoogleBookConnectionView::class.java)
    }
}